clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java

```diff
@@ -500,7 +508,19 @@ public Record next() {
                 throw new NoSuchElementException();
 
             readRecords++;
-            return readNext(baseOffset, baseTimestamp, baseSequence, logAppendTime);
+            Record rec = readNext(baseOffset, baseTimestamp, baseSequence, logAppendTime);
+            if (readRecords == numRecords) {
+                // validate the batch
+                boolean isValid = false;
+                try {
+                    readNext(baseOffset, baseTimestamp, baseSequence, logAppendTime);
```

Comment:
One issue is that `available` doesn't have the contract we want. See the following comment for more details.

https://github.com/apache/kafka/pull/2025#issuecomment-260653068